cli/explain/compactgraph/input.go

```diff
@@ -509,12 +531,24 @@ func protoToRunfilesTree(r *spawn.ExecLogEntry_RunfilesTree, previousInputs map[
 		contentHash.Write(repoMappingManifest.ShallowContentHash())
 	}
 
+	binary.Write(contentHash, binary.LittleEndian, uint64(len(r.EmptyFiles)))
+	internedEmptyFiles := make([]string, 0, len(r.EmptyFiles))
+	for _, emptyFilePath := range r.EmptyFiles {
+		binary.Write(contentHash, binary.LittleEndian, uint64(len(emptyFilePath)))
+		contentHash.Write([]byte(emptyFilePath))
+		// Bazel emits an empty file for each parent of a directory with a Python file in it. This typically results in
+		// many duplicated paths across Python runfiles trees. Interning them reduces the retained memory to be roughly
+		// linear in the compressed size of the exec log.
+		internedEmptyFiles = append(internedEmptyFiles, interner(emptyFilePath))
+	}
```

Comment:
Im a bit confused here: we are writing to the content hash the empty file count, then the length of each file path and the file paths? Is it to differentiate against some runfiletrees which may have the last input that looks exactly like a list of emptyfiles?

I wonder if we just need a "magic" separator added to this Hash 🤔

Reply:
The aim is to keep the digested bytes parseable, meaning that it would be possible to reconstruct the data structure from the byte stream, which guarantees that there can't be collisions, for example between inputs and empty files or within a single empty file path (that looks like two other concatenated). Since file paths are themselves essentially arbitrary streams of bytes of variable length, static separators won't help and length encoding is decently simple.